File: scripts/compare_datasets.R
#!/usr/bin/env Rscript
# R side of the real-dataset comparison. Fits glmnet on each dataset and writes
# the lambda path, coefficients, and fit time so the Python side can compare
# coefficients and wall clock. Reads the same CSVs the Python side reads.

suppressMessages(library(glmnet))
suppressMessages(library(jsonlite))

here <- dirname(sub("^--file=", "", grep("^--file=", commandArgs(FALSE), value = TRUE)[1]))
data_dir <- file.path(here, "..", "datasets")
out_dir <- file.path(here, "..", "datasets", ".compare")
dir.create(out_dir, showWarnings = FALSE, recursive = TRUE)

time_fit <- function(fun, reps = 5) {
  fun() # warmup
  best <- Inf
  for (i in seq_len(reps)) {
    t0 <- proc.time()[["elapsed"]]
    fit <- fun()
    best <- min(best, proc.time()[["elapsed"]] - t0)
  }
  list(fit = fun(), time = best)
}

dump <- function(name, fit, time) {
  write(toJSON(list(
    lmu = length(fit$lambda),
    lambda = as.vector(fit$lambda),
    a0 = as.vector(fit$a0),
    beta = as.vector(as.matrix(fit$beta)), # p x lmu, column-major
    p = nrow(fit$beta),
    time = time,
    npasses = fit$npasses
  ), digits = NA, auto_unbox = TRUE), file.path(out_dir, paste0(name, ".json")))
}

# --- wine quality (long, gaussian) -----------------------------------------
w <- read.csv(file.path(data_dir, "winequality-white.csv"), sep = ";")
Xw <- as.matrix(w[, setdiff(names(w), "quality")])
yw <- w$quality
rw <- time_fit(function() glmnet(Xw, yw, family = "gaussian", type.gaussian = "naive"))
dump("wine", rw$fit, rw$time)
cat(sprintf("wine     n=%d p=%d  lmu=%d  R fit=%.1f ms\n",
            nrow(Xw), ncol(Xw), length(rw$fit$lambda), rw$time * 1e3))

# --- leukemia (wide, binomial) ---------------------------------------------
lab <- as.character(read.csv(file.path(data_dir, "leukemia_big.csv"), header = FALSE, nrows = 1))
expr <- as.matrix(read.csv(file.path(data_dir, "leukemia_big.csv"), header = FALSE, skip = 1))
Xl <- t(expr) # samples x genes
yl <- as.integer(lab == "AML")
rl <- time_fit(function() glmnet(Xl, yl, family = "binomial", type.logistic = "Newton"))
dump("leukemia", rl$fit, rl$time)
cat(sprintf("leukemia n=%d p=%d  lmu=%d  R fit=%.1f ms\n",
            nrow(Xl), ncol(Xl), length(rl$fit$lambda), rl$time * 1e3))

File: scripts/compare_datasets.py
#!/usr/bin/env python
"""Compare rust-glmnet against R glmnet on two real datasets -- one long
(n >> p) and one wide (p >> n) -- on both coefficients and wall clock.

Both sides read the same CSVs (see datasets/README.md) and fit the full lambda
path with matching solver settings, so the comparison is apples-to-apples. The R
side (scripts/compare_datasets.R) writes its results to datasets/.compare/;
run it first, or let this script invoke it.
"""

import json
import pathlib
import subprocess
import sys
import time

import numpy as np
from glmnetrs import glmnet

HERE = pathlib.Path(__file__).parent
DATA = HERE / ".." / "datasets"
COMPARE = DATA / ".compare"


def load_wine():
    """Long, gaussian: predict wine quality (n=4898, p=11)."""
    arr = np.genfromtxt(DATA / "winequality-white.csv", delimiter=";", skip_header=1)
    return arr[:, :-1], arr[:, -1], "gaussian"


def load_leukemia():
    """Wide, binomial: ALL vs AML from gene expression (n=72, p=7128)."""
    with open(DATA / "leukemia_big.csv") as fh:
        labels = np.array(fh.readline().strip().split(","))
    expr = np.loadtxt(DATA / "leukemia_big.csv", delimiter=",", skiprows=1)
    return expr.T, (labels == "AML").astype(float), "binomial"


def time_ours(X, y, family, reps=5):
    glmnet(X, y, family=family)  # warmup
    best = np.inf
    fit = None
    for _ in range(reps):
        t0 = time.perf_counter()
        fit = glmnet(X, y, family=family)
        best = min(best, time.perf_counter() - t0)
    return fit, best


def compare(name, fit, ours_time):
    r = json.loads((COMPARE / f"{name}.json").read_text())
    p = r["p"]
    lmu = min(fit.lmu, r["lmu"])
    r_beta = np.asarray(r["beta"]).reshape((p, r["lmu"]), order="F")
    r_a0 = np.asarray(r["a0"])
    r_lam = np.asarray(r["lambda"])

    def rel(a, b):
        return np.max(np.abs(a - b) / (1.0 + np.abs(b)))

    lam_err = rel(fit.lambda_[:lmu], r_lam[:lmu])
    a0_err = rel(fit.a0[:lmu], r_a0[:lmu])
    beta_err = rel(fit.beta[:, :lmu], r_beta[:, :lmu])
    return {
        "name": name,
        "p": p,
        "lmu_ours": fit.lmu,
        "lmu_r": r["lmu"],
        "lam_err": lam_err,
        "a0_err": a0_err,
        "beta_err": beta_err,
        "ours_ms": ours_time * 1e3,
        "r_ms": r["time"] * 1e3,
    }


def main():
    if not COMPARE.exists() or not (COMPARE / "wine.json").exists():
        print("Running R side (scripts/compare_datasets.R) ...", flush=True)
        rc = subprocess.run(["Rscript", str(HERE / "compare_datasets.R")])
        if rc.returncode != 0:
            print("R side failed; run it manually.", file=sys.stderr)
            sys.exit(1)

    rows = []
    for name, loader in (("wine", load_wine), ("leukemia", load_leukemia)):
        X, y, family = loader()
        fit, t = time_ours(X, y, family)
        rows.append(compare(name, fit, t))

    print("\n" + "=" * 86)
    print(
        f"{'dataset':10} {'n':>5} {'p':>6} {'lmu':>4} "
        f"{'max coef relerr':>16} {'ours':>9} {'R':>9} {'speedup':>8}"
    )
    print("-" * 86)
    for r, (name, loader) in zip(rows, (("wine", load_wine), ("leukemia", load_leukemia))):
        X, _, _ = loader()
        worst = max(r["lam_err"], r["a0_err"], r["beta_err"])
        speed = r["r_ms"] / r["ours_ms"]
        lmu = f"{r['lmu_ours']}" if r["lmu_ours"] == r["lmu_r"] else f"{r['lmu_ours']}/{r['lmu_r']}"
        print(
            f"{name:10} {X.shape[0]:>5} {X.shape[1]:>6} {lmu:>4} "
            f"{worst:>16.2e} {r['ours_ms']:>7.1f}ms {r['r_ms']:>7.1f}ms {speed:>7.2f}x"
        )
    print("=" * 86)
    print("max coef relerr: worst relative difference in lambda / intercept / beta vs R.")


if __name__ == "__main__":
    main()
